### PR TITLE
Ask all the questions about identity in the beginning and validate

### PR DESCRIPTION
**Pull request by grimpy**
against explorer

Also added a question to pick your explorer

Signed-off-by: Jo De Boeck <deboeck.jo@gmail.com>

### CODE REVIEW

**Review by xmonader — APPROVED** [review]

**Review by waleedhammam — APPROVED** [review]
Thanks ! very sleek

File: install/threesdk/lib/SDKContainers.py [hunk 1]
```diff
@@ -6,20 +14,78 @@ def __init__(self, core, args):
...
-        if "." not in identity:
-            identity += ".3bot"
-        identity = identity.lower()
-        if self.args.identity != identity:
-            self.args.identity = identity
-            self.args.words = None
-            self.args.email = None
-
-        return identity
+    def _get_user(self):
```
Comment (zaibon) [inline install/threesdk/lib/SDKContainers.py:17]: keep in mind that email are also unique.

File: install/threesdk/lib/SDKContainers.py [hunk 2]
```diff
@@ -6,20 +14,78 @@ def __init__(self, core, args):
...
+        else:
+            users = resp.json()
+            if users:
+                return users[0]
+            return None
+
+    def _identity_ask(self, identity=None, explorer=None):
+        def _fill_identity_args(identity, explorer):
+            def fill_words():
+                words = self.core.IT.Tools.ask_string("what are the words associated with your identity?")
```
Comment (zaibon) [inline install/threesdk/lib/SDKContainers.py:30]: shouldn't we more explicit and says, copy the words from your 3bot connect app ?
Also for the threebot name, it should be the same as in the 3bot app.

File: install/threesdk/lib/SDKContainers.py [hunk 3]
```diff
@@ -6,20 +14,78 @@ def __init__(self, core, args):
...
+                    fill_words()
+                else:
+                    self.args.words = None
+            if user and self.args.words:
+                # time to do validation of words
+                while True:
+                    try:
+                        seed = self.core.IT.Tools.to_entropy(self.args.words, english.words)
+                    except Exception:
+                        choice = self.core.IT.Tools.ask_choices("Invalid words where given what would you like to do?", ["restart", "reenter"])
```
Comment (zaibon) [inline install/threesdk/lib/SDKContainers.py:70]: Should be more explicit about the error. What does invalid means ? Why ? and what is the advice to get it right ?

File: install/threesdk/lib/SDKContainers.py [hunk 4]
```diff
@@ -6,20 +14,78 @@ def __init__(self, core, args):
...
+                        if choice == "restart":
+                            return False
+                        fill_words()
+                        continue
+                    # new we have a valid seed let's check if it matches the user
+                    key = SigningKey(seed)
+                    hexkey = binascii.hexlify(key.verify_key.encode()).decode()
+                    if hexkey != user["pubkey"]:
+                        choice = self.core.IT.Tools.ask_choices("Your words do not match your idenitiy, what would you like to do?", ["restart", "reenter"])
```
Comment (zaibon) [inline install/threesdk/lib/SDKContainers.py:79]: better would be: "There seems to be an identity registered with the same name and/or email address but a different public key."
No this brings the question, what do we do when user have lost their seed ? We have to remove them from the phonebook manually ?

**Review by zaibon — CHANGES_REQUESTED** [review]

**Comment by zaibon** [comment]
Can you also make sure to also create a PR that update the documentation for install and the manual

File: install/threesdk/lib/SDKContainers.py (hunk 2 above)
Comment (rkhamis) [reply install/threesdk/lib/SDKContainers.py:30]: Let's make this:

"Copy the phrase from your 3bot Connect app here"

File: install/threesdk/lib/SDKContainers.py [hunk 5]
```diff
@@ -6,20 +14,78 @@ def __init__(self, core, args):
...
+            if not self.args.explorer:
+                return True
+            if not self.args.identity:
+                identity = self.core.IT.Tools.ask_string("what is your threebot name (identity)?")
+                if "." not in identity:
+                    identity += ".3bot"
+                self.args.identity = identity
+            user = self._get_user()
+            if not user and not self.args.email:
+                email = self.core.IT.Tools.ask_string("what is your email associated with your identity?")
```
Comment (rkhamis) [inline install/threesdk/lib/SDKContainers.py:54]: *What is the email address associated with your identity

File: install/threesdk/lib/SDKContainers.py (hunk 3 above)
Comment (rkhamis) [reply install/threesdk/lib/SDKContainers.py:70]: here we can have the message say:

"Seems one or more more words entered is invalid. What would you like to do?" with 
"restart initialization of identity" and "reenter phrase" as options

File: install/threesdk/lib/SDKContainers.py (hunk 4 above)
Comment (rkhamis) [reply install/threesdk/lib/SDKContainers.py:79]: so far we have no way other than manually indeed

**Review by rkhamis — CHANGES_REQUESTED** [review]

File: install/threesdk/lib/SDKContainers.py (hunk 4 above)
Comment (zaibon) [reply install/threesdk/lib/SDKContainers.py:79]: We should define a process where people can ping us to ask for a removal. Of course we need to have a way to make sure they are who they claim to be.... which is the tricky part...

File: install/threesdk/lib/SDKContainers.py [hunk 6]
```diff
@@ -49,34 +56,50 @@ def fill_words():
...
                 print("Configured email for this identity is {}".format(user["email"]))
                 self.args.email = user["email"]
+
             if not self.args.words:
-                if user:
-                    fill_words()
-                else:
-                    self.args.words = None
-            if user and self.args.words:
+                fill_words()
```
Comment (grimpy) [inline install/threesdk/lib/SDKContainers.py:78]: i was making it possible to leave words empty if it's a new user. to they can be generated later on by container install.. now you are forcing the user to enter words...

File: install/setup.py [hunk 7]
```diff
@@ -8,7 +8,7 @@
     author="Threefold Tech",
     author_email="info@threefold.tech",
     url="http://github.com/threefoldtech/jumpscaleX_core",
-    install_requires=["requests", "ptpython==2.0.4", "pudb", "jedi"],
+    install_requires=["requests", "ptpython==2.0.4", "pudb", "jedi", "nacl"],
```
Comment (grimpy) [inline install/setup.py:11]: this should also become pynacl

File: install/threesdk/lib/SDKContainers.py (hunk 6 above)
Comment (waleedhammam) [reply install/threesdk/lib/SDKContainers.py:78]: yeah but that caused a recursion for me

File: install/threesdk/lib/SDKContainers.py [hunk 8]
```diff
@@ -6,20 +14,101 @@ def __init__(self, core, args):
...
+                            ["restart", "reenter"],
+                        )
+                        if choice == "restart":
+                            return False
+                        fill_words()
+                        continue
+                    # new we have a valid seed let's check if it matches the user
+                    key = SigningKey(seed)
+                    hexkey = binascii.hexlify(key.verify_key.encode()).decode()
+                    if user and hexkey != user["pubkey"]:
```
Comment (waleedhammam) [inline install/threesdk/lib/SDKContainers.py:97]: I tried with my old identity yet it didn't success

**Pull request merged**